Changes in module 'nf-core/bwa/mem'
--- modules/nf-core/bwa/mem/main.nf
+++ modules/nf-core/bwa/mem/main.nf
@@ -13,8 +13,8 @@
     val   sort_bam
 
     output:
-    tuple val(meta), path("*.bam"), emit: bam
-    path  "versions.yml"          , emit: versions
+    tuple val(meta), path("*.cram"),    emit: cram
+    path  "versions.yml",               emit: versions
 
     when:
     task.ext.when == null || task.ext.when
@@ -30,9 +30,10 @@
     bwa mem \\
         $args \\
         -t $task.cpus \\
+        -R "@RG\\tID:$prefix\\tSM:bar" \\
         \$INDEX \\
         $reads \\
-        | samtools $samtools_command $args2 --threads $task.cpus -o ${prefix}.bam -
+        | samtools $samtools_command $args2 --threads $task.cpus -O cram -o ${prefix}.cram -
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
@@ -44,7 +45,7 @@
     stub:
     def prefix = task.ext.prefix ?: "${meta.id}"
     """
-    touch ${prefix}.bam
+    touch ${prefix}.cram
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
