Changes in module 'nf-core/samtools/depth'
--- modules/nf-core/samtools/depth/main.nf
+++ modules/nf-core/samtools/depth/main.nf
@@ -9,11 +9,13 @@
 
     input:
     tuple val(meta1), path(bam)
-    tuple val(meta2), path(intervals)
+    tuple val(meta2), path(bai)
+    tuple val(meta3), val(chromosome)
 
     output:
-    tuple val(meta1), path("*.tsv"), emit: tsv
-    path "versions.yml"           , emit: versions
+    tuple val(meta3), path("*.tsv.gz"),     emit: depth
+    tuple val(meta1), path("*.list.txt"),   emit: bam_list
+    path "versions.yml",                    emit: versions
 
     when:
     task.ext.when == null || task.ext.when
@@ -21,15 +23,18 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta1.id}"
-    def positions = intervals ? "-b ${intervals}" : ""
+    def positions = chromosome ? "-r ${chromosome}" : ""
     """
+    ls $bam | xargs -n1 > ${prefix}.list.txt
+
     samtools \\
         depth \\
         --threads ${task.cpus-1} \\
         $args \\
         $positions \\
-        -o ${prefix}.tsv \\
-        $bam
+        -H -a \\
+        -f ${prefix}.list.txt \\
+        | gzip -c > ${prefix}.${meta3.id}.depth.tsv.gz
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
